internal/lint/lint.go

```diff
@@ -0,0 +1,335 @@
+	for _, checker := range AllCheckers {
+		if _, ok := ids[checker.ID()]; ok {
+			panic(fmt.Sprintf("duplicate checker id %s", checker.ID()))
+		}
+		ids[checker.ID()] = struct{}{}
+	}
+}
+
+// Runner runs a lint job.
+type Runner interface {
+	Run(...*file.ProtoSet) ([]*text.Failure, error)
+}
+
+// RunnerOption is an option for a new Runner.
+type RunnerOption func(*runner)
+
+// RunnerWithLogger returns a RunnerOption that uses the given logger.
+//
+// The default is to use zap.NewNop().
+func RunnerWithLogger(logger *zap.Logger) RunnerOption {
+	return func(runner *runner) {
+		runner.logger = logger
+	}
+}
+
+// NewRunner returns a new Runner.
+func NewRunner(options ...RunnerOption) Runner {
+	return newRunner(options...)
+}
+
+// The below should not be needed in the CLI
+// TODO make private
+
+// Checker is a linter for Protobuf files.
+type Checker interface {
+	// Return the ID of this Checker. This should be all UPPER_SNAKE_CASE.
+	ID() string
+	// Return the purpose of this Checker. This should be a human-readable string.
+	Purpose() string
+	// Check the file data for the descriptors in a common directgory.
+	// If there is a lint failure, this returns it in the
+	// slice and does not return an error. An error is returned if something
+	// unexpected happens. Callers should verify the files are compilable
+	// before running this.
+	Check(dirPath string, descriptors []*proto.Proto) ([]*text.Failure, error)
+}
+
+// NewChecker is a convenience function that returns a new Checker for the
+// given parameters.
+//
+// The ID will be upper-cased.
+//
+// Failures returned from check do not need to set the ID, this will be overwritten.
+func NewChecker(id string, purpose string, check func(string, []*proto.Proto) ([]*text.Failure, error)) Checker {
+	return newBaseChecker(id, purpose, check)
+}
+
+// NewAddChecker is a convenience function that returns a new Checker for the
+// given parameters, using a function to record failures.
+//
+// The ID will be upper-cased.
+//
+// Failures returned from check do not need to set the ID, this will be overwritten.
+func NewAddChecker(id string, purpose string, addCheck func(func(*text.Failure), string, []*proto.Proto) error) Checker {
+	return newBaseAddChecker(id, purpose, addCheck)
+}
+
+// GetCheckers returns the Checkers for the LintConfig.
+//
+// The config is expected to be valid, ie slices deduped, all upper-case,
+// and only either IDs or Group/IncludeIDs/ExcludeIDs, with no overlap between
+// IncludeIDs and ExcludeIDs.
+//
+// If the config came from the settings package, this is already validated.
+func GetCheckers(config settings.LintConfig) ([]Checker, error) {
+	if len(config.IDs) == 0 && (len(config.Group) == 0 || config.Group == DefaultGroup) && len(config.IncludeIDs) == 0 && len(config.ExcludeIDs) == 0 {
+		return DefaultCheckers, nil
+	}
+
+	if len(config.IDs) > 0 {
+		var checkers []Checker
+		// n^2 woot
+		for _, checker := range AllCheckers {
```

Comment:
We could also make `AllCheckers` a map with ID keys to eliminate the iteration.
```
if checker, ok := AllCheckers[id /* config.ID */]; ok {
  checkers = append(checkers, checker)
}
```

Also, should we be returning an error if an ID is provided that we don't recognize?

Reply:
Not sure if I follow here. Would we just need to range over the `config.IDs` and check to see if the key exists?

```
var checkers []Checker
for _, id := range config.IDs {
  if checker, ok := AllCheckers[id]; ok {
    checkers = append(checkers, checker)
  }
}
```

Reply:
I see the confusion here.
@amckinney, `AllCheckers` is initialized [here](https://sourcegraph.com/github.com/uber/prototool@d8678e3fe743bede4a7ee98bd9d15f396517fc81/-/blob/internal/lint/lint.go#L37:2) so we can't create that mapping in the same way that we're creating that slice.
@peter-edge what if we create a new map (eg `allCheckers[id]Checker`) before these calls, then we can reduce the n^2 to n

Reply:
We could. Just want to point out that this takes on the order of microseconds (I haven't benchmarked it, but I can almost guarantee that lol), the number of checkers is not going to explode to be 1,000,000 checkers. I rather keep it simple personally, but I can change this if you want.

Reply:
Yeah true, i'm indifferent. @amckinney your call, since you brought it up

Reply:
Got it, yea this is fine as is. :+1: